src/Npgsql/Internal/Converters/Primitive/ByteaConverters.cs

```diff
@@ -101,24 +101,55 @@ public override ValueTask<Stream> ReadAsync(PgReader reader, CancellationToken c
 
     public override Size GetSize(SizeContext context, Stream value, ref object? writeState)
     {
-        var memoryStream = new MemoryStream(value.CanSeek ? (int)(value.Length - value.Position) : 0);
+        if (value.CanSeek)
+            return checked((int)(value.Length - value.Position));
+
+        var memoryStream = new MemoryStream();
         value.CopyTo(memoryStream);
         writeState = memoryStream;
         return checked((int)memoryStream.Length);
     }
 
     public override void Write(PgWriter writer, Stream value)
     {
-        if (!((MemoryStream)writer.Current.WriteState!).TryGetBuffer(out var segment))
-            throw new InvalidOperationException();
-        writer.WriteBytes(segment.AsSpan());
+        if (writer.Current.WriteState is not null)
+        {
+            if (!((MemoryStream)writer.Current.WriteState!).TryGetBuffer(out var writeStateSegment))
+                throw new InvalidOperationException();
+
+            writer.WriteBytes(writeStateSegment.AsSpan());
+            return;
+        }
+
+        // Non-derived MemoryStream fast path
+        if (value.GetType() == typeof(MemoryStream) && ((MemoryStream)value).TryGetBuffer(out var segment))
+            writer.WriteBytes(segment.AsSpan((int)value.Position));
+        else
+            value.CopyTo(writer.GetStream(allowMixedIO: true));
```

Comment:
No need to pass allowMixedIO. As opposed to the NetTopologySuiteConverter we have both Write and WriteAsync implementations here.